pyscriptjs/src/main.ts

```diff
@@ -1,31 +1,103 @@
 import './styles/pyscript_base.css';
 
+import { loadConfigFromElement } from './pyconfig';
+import type { AppConfig } from './pyconfig';
+import type { Runtime } from './runtime';
 import { PyScript } from './components/pyscript';
 import { PyEnv } from './components/pyenv';
 import { PyLoader } from './components/pyloader';
-import { PyConfig } from './components/pyconfig';
+import { PyodideRuntime } from './pyodide';
 import { getLogger } from './logger';
-import { globalLoader } from './stores';
+import { globalLoader, runtimeLoaded, addInitializer } from './stores';
+import { handleFetchError, globalExport } from './utils'
 
 const logger = getLogger('pyscript/main');
 
-/* eslint-disable @typescript-eslint/no-unused-vars */
-const xPyScript = customElements.define('py-script', PyScript);
-const xPyLoader = customElements.define('py-loader', PyLoader);
-const xPyConfig = customElements.define('py-config', PyConfig);
-const xPyEnv = customElements.define('py-env', PyEnv);
-/* eslint-disable @typescript-eslint/no-unused-vars */
-
-// As first thing, loop for application configs
-logger.info('checking for py-config');
-const config: PyConfig = document.querySelector('py-config');
-if (!config) {
-    const loader = document.createElement('py-config');
-    document.body.append(loader);
+// XXX this should be killed eventually
+let runtimeSpec: Runtime;
+runtimeLoaded.subscribe(value => {
+    runtimeSpec = value;
+});
+
+
+class PyScriptApp {
+
+    config: AppConfig;
+
+    main() {
+        this.loadConfig();
+        this.initialize();
+
+        /* eslint-disable @typescript-eslint/no-unused-vars */
+        const xPyScript = customElements.define('py-script', PyScript);
+        const xPyLoader = customElements.define('py-loader', PyLoader);
+        const xPyEnv = customElements.define('py-env', PyEnv);
+        /* eslint-disable @typescript-eslint/no-unused-vars */
+
+        // add loader to the page body
+        logger.info('add py-loader');
+        const loader = <PyLoader>document.createElement('py-loader');
+        document.body.append(loader);
+        globalLoader.set(loader);
+    }
+
+    loadConfig() {
+        // find the <py-config> tag. If not found, we get null which means
+        // "use the default config"
+        // XXX: what happens if we have multiple ones?
+        logger.info('searching for <py-config>');
+        const el = document.querySelector('py-config');
+        this.config = loadConfigFromElement(el);
+        logger.info('config loaded:\n' + JSON.stringify(this.config, null, 2));
+    }
+
+    initialize() {
+        addInitializer(this.loadPackages);
+        addInitializer(this.loadPaths);
+        this.loadRuntimes();
+    }
+
+    loadPackages = async () => {
+        logger.info("Packages to install: ", this.config.packages);
+        await runtimeSpec.installPackage(this.config.packages);
```

Comment:
same answer as above: the messy logic didn't really change here, so it might be possible. But I want to refactor it in a way that `runtime` is always passed as a parameter so that when you receive you are sure that it's ready

Reply:
That makes sense I missed that context thank you!